### PR TITLE
PCSX2-WX: Add proper vertical frequency values on GS Frame

### PR DESCRIPTION
**Pull request by ssakash**
**Summary of changes**:

* Add proper speed percentage calculation for 480P video mode. ``(60 -> 59.94)``

### CODE REVIEW

File: pcsx2/Counters.cpp [hunk 1]
```diff
@@ -270,6 +270,30 @@ static const char* ReportVideoMode()
...
+	case GS_VideoMode::Uninitialized: // SYSCALL instruction hasn't executed yet, give some temporary values.
+		return 60;
+	case GS_VideoMode::PAL:
+		return EmuConfig.GS.FrameratePAL;
+	case GS_VideoMode::NTSC:
+		return EmuConfig.GS.FramerateNTSC;
+	case GS_VideoMode::HDTV_480P:
+		return 59.94;
+	case GS_VideoMode::HDTV_1080P:
+	case GS_VideoMode::HDTV_1080I:
```
Comment (refractionpcsx2) [inline pcsx2/Counters.cpp]: I thought 1080i was 30fps? PAL will be, surely?

Comment (ssakash) [inline pcsx2/Counters.cpp]: > I thought 1080i was 30fps?

Vertical frequency is the refresh rate controlled by the vertical blanking signal of PCRTC circuit, not related to fps. The actual frames pushed are 30fps as you had said earlier. (half of vertical frequency)

> PAL will be, surely?

What do you mean ? :P

Comment (refractionpcsx2) [inline pcsx2/Counters.cpp]: Sorry I meant Progressive, but if that's what you mean then that will probably be 60 too.

Comment (ssakash) [inline pcsx2/Counters.cpp]: Since today is a special day, I'll explain it more thoroughly :P 

There are generally two types of frequency for a video mode, one is vertical and the other one is horizontal.

```
----.-----------.---------------.-----------.-----------.
 No | Name      | Resolution	| fV(Hz)    | fH(kHz)   |
----+-----------+---------------+-----------+-----------|
  0 | NTSC-NI   |  640x240(224) | 59.940    | 15.734    |
  1 | NTSC-I    |  640x480(448) | 59.820    | 15.734    |
  2 | PAL-NI    |  640x288(256) | 50.000    | 15.625    |
  3 | PAL-I     |  640x576(512) | 49.760    | 15.625    |
  4 | VESA-1A   |  640x480      | 59.940    | 31.469    |
  5 | VESA-1C   |  640x480      | 75.000    | 37.500    |
  6 | VESA-2B   |  800x600      | 60.317    | 37.879    |
  7 | VESA-2D   |  800x600      | 75.000    | 46.875    |
  8 | VESA-3B   | 1024x768      | 60.004    | 48.363    |
  9 | VESA-3D   | 1024x768      | 75.029    | 60.023    |
 10 | VESA-4A   | 1280x1024     | 60.020    | 63.981    |
 11 | VESA-4B   | 1280x1024     | 75.025    | 79.976    |
 12 | DTV-480P  |  720x480      | 59.940    | 31.469    |
 13 | DTV-1080I | 1920x1080     | 60.000    | 33.750    |
 14 | DTV-720P  | 1280x720      | ??        | ??        |
----^-----------^---------------^-----------^-----------'
```

Note that vertical frequency is _different_ from the [horizontal frequency](https://en.wikipedia.org/wiki/Horizontal_scan_rate) and frame frequency. (AKA FPS)

Here's an example from Wikipedia : 
> For example, most movie projectors advance from one frame to the next one 24 times each second. But each frame is illuminated two or three times before the next frame is projected using a shutter in front of its lamp. As a result, the movie projector runs at 24 frames per second, but has a 48 or 72 Hz refresh rate.

So in our case,

NTSC -> Vertical frequency of 59.94Hz and a frame frequency of 29.97 fps
PAL -> Vertical frequency of 50 Hz and a frame frequency of 25 fps
HDTV 480p -> Vertical frequency of 59.94Hz and a frame frequency of 29.97 fps
Other progressive modes ->  Vertical frequency of 60Hz and a frame frequency of 30 fps

Also progressive mode isn't a specific video mode, it's just a term used to describe all the video modes which have not set the INT flag.

Comment (refractionpcsx2) [inline pcsx2/Counters.cpp]: Thanks for the patronising response :P The table would have done perfectly lol

**Comment by avih** [comment]
Do you know of games or emulation states which would be affected by it in practice? it would seem the bios would too, right? If so, shouldn't all those "60" be derived from from the NTSC configuration value instead (i.e. that or half of it, whichever PCSX2 needs)?

Also, assuming I'm reading it correctly and it would set "60" while at the bios, does it mean a real PS2, regardless if PAL or NTSC is pushing exactly 60 FPS to thee TV set while booting? not 59.94 or 50?

(Yes, I know the difference between FPS and vertical frequency and interlacing).

**Comment by ssakash** [comment]
> Do you know of games or emulation states which would be affected by it in practice?

This commit doesn't affect the internal counters system. The commit only influences the Vsync rate percentage calculation on GS frame, the internal vertical frequency and frame frequency calculation is still the same as master branch.

> Also, assuming I'm reading it correctly and it would set "60" while at the bios, does it mean a real PS2, regardless if PAL or NTSC is pushing exactly 60 FPS to the TV set while booting? not 59.94 or 50 (or half of those, depending how you count)?

Not relevant to the current PR (Related to **#1442**). Either way, I'll still explain. Previously there was a certain bug in the counters code, whenever we load any medium we will always initialize the video mode with NTSC timing values which was wrong and not _accurate_ to the PS2 behavior.

I removed that early init code and initialized the vertical frequency / frame rate values based on the video mode identified via SetGsCrt which is the PS2 behavior.

Also I do have to accept that the current frame frequency calculation code for bios is partially inaccurate. Sony's binary loader which does [regional checkout](https://en.wikipedia.org/wiki/Regional_lockout) uses regular PAL/NTSC DVD video mode (**not** the regular PAL/NTSC video mode) which has a frame frequency of 25/30. I only had a USA bios which uses NTSC DVD video mode format so I set it it's frequency globally for all bios video modes. (I know, I'm wrong...)

The correct behavior is to set 25 * 2 = 50Hz when detecting an Europe bios, USA bios behavior is correct. I don't have an Europe bios to check it out so someone else has to do it. Either way, it really isn't picky on the frequency so I don't really care. The older code was much worse (automatic NTSC initialization), though it was still fine.

**Comment by avih** [comment]
Appreciate the explanation.

So in practice, it would affect the frequency which we consider as base when calculating the speed percentage for the title (regardless of whether or not the other value which we compare it to is accurate or not), but would not affect anything outside the title display? for instance no games would run faster or slower due to this?

Just to get a better understanding, and I appreciate your patience, what happens if we get it wrong? for instance if it should be 50 but we're accidentally setting this to 60 when the user wants to run in normal speed (no turbo, no slowmo, etc)? the title would show 83% and that's where it ends? otherwise it still runs at the same speed?

**Comment by ssakash** [comment]
> So in practice, it would affect the frequency which we consider as base when calculating the speed percentage for the title (regardless of whether or not the other value which we compare it to is accurate or not), but would not affect anything outside the title display? for instance no games would run faster or slower due to this?

Yes, the change only influences the calculation of percentage value on GS frame.

> what happens if we get it wrong? for instance if it should be 50 but we're accidentally setting this to 60 when the user wants to run in normal speed (no turbo, no slowmo, etc)? the title would show 83% and that's where it ends? otherwise it still runs at the same speed?

If we got it wrong then only the percentage calculation would be wrong, the internal speed will still be the same.

**Comment by avih** [comment]
How is it better than what we had in the past? As far as I recall it was displaying the speed correctly. Or was i just lucky to never test cases where it was wrong? Do you know such cases which improve due to this? Care to give an example?

**Comment by ssakash** [comment]
> Do you know such cases which improve due to this? Care to give an example?

The only case where it notices an improvement from an accuracy perspective is on 480p video mode. Quoting my commit message,

```
Previous code considered 60 FPS as limit for all progressive modes, however 480p uses 59.94....
Yes, I know. I'm being too nitpicky.
```

**Comment by avih** [comment]
So for instance GT4 in 480p more would show on average 100.1% instead of 100% when it runs at full speed?

**Comment by avih** [comment]
Wait a second, isn't this commit pure refactor with zero functional changes? I just tried to follow the values, and as far as I can tell they're always identical to how they were before.

Because you refactored and (maybe?) updated one value, it makes it really hard to tell what actually got changed.

If you claim that 480 was changed, then I think it was not. It was 29.97, and now it's 59.94/2... Am I missing something? And elsewhere I also can't notice any changes... other than refactoring the code... Please tell me I'm wrong.

**Comment by ssakash** [comment]
> So for instance GT4 in 480p would show on average 100.1% instead of 100% when it runs at full speed?

Yes, 100.001% increase on the GS frame to be precise. Internal speed of the game will still remain same as master branch.

> If you claim that 480 was changed, then I think it was not. It was 29.97, and now it's 59.94/2... Am I missing something?

``29.97`` and ``59.94/2`` are both same.

**Comment by avih** [comment]
So this is pure refactor with zero effect on the tile display or anything else, right?

**Comment by ssakash** [comment]
Please look at the [following line](https://github.com/PCSX2/pcsx2/pull/1491/files#diff-d32bc35f2098a256f429f22a1c90d129R594) , it does **influence** the percentage calculation on title bar.

**Comment by ssakash** [comment]
This was the older code

```
gsVideoMode == GS_VideoMode::PAL ? (fps * 100) / EmuConfig.GS.FrameratePAL.ToFloat() : (fps * 100) / EmuConfig.GS.FramerateNTSC.ToFloat();
```

The [new change](https://github.com/PCSX2/pcsx2/pull/1491/files#diff-d32bc35f2098a256f429f22a1c90d129R594) passes the vertical frequency values of individual video modes from the counters to the GUI.

**Comment by avih** [comment]
> Please look at the following line , it does influence the percentage calculation on title bar.

Right, now I finally get it. At which case, please split this commit into two, at one do the refactor and mark it as no-op (for instance `Refactor GS vertical frequency (no-op)`, and at the other change that one line which actually changes the calculation for the title, and use an appropriate commit message.

It's really hard to follow a commit when it has 30 lines of no-op refactor and only one line which actually changed.

**Comment by avih** [comment]
(also, not that I finally get it, I do like both the refactor and the better calculation, just please make it easier for others to follow and understand ;) )

**Comment by avih** [comment]
Slightly off topic, it seems like we already have NTSC and PAL overrides at the config, but the user can't override the speed for progressive display modes (AFAIK 480p is the only one the PS2 supports, right? GT4 can do 1080i, which would need to use the NTSC or possibly PAL values I guess).

So it would probably nice to have PROGRESSIVE speed override too, for instance, I could easily see users wanting progressive games to display at 59.94 instead of at 60 because most PC monitors are actually 59.94 and not 60, so possibly smoother playback experience.

**Comment by ssakash** [comment]
> AFAIK 480p is the only one the PS2 supports, right?

Technically, PS2 is also capable of 576p , 720p and 1080p however no commercial games use it. so yeah, 480p is the only _official_ progressive mode used by the PS2 but however do note that there are also other cases where INT flag is not set when using PAL/NTSC video mode which is also progressive mode. ( **Example**: ICO)

> So it would probably nice to have PROGRESSIVE speed override too, for instance, I could easily see users wanting progressive games to display at 59.94 instead of at 60 because most PC monitors are actually 59.94 and not 60, so possibly smoother playback experience.

Previously on https://github.com/PCSX2/pcsx2/pull/1145#issuecomment-176208735 you had suggested to remove the customization boxes from the GUI for non-development builds. Guess you had a change of mind ? Either way, I don't really have any opinion regarding addition of a progressive mode speed override.

**Comment by avih** [comment]
> Technically, PS2 is also capable of 576p ...

Good info, thanks.

> Previously on #1145 (comment) you had suggested to remove the customization boxes from the GUI for non-development builds. Guess you had a change of mind ? Either way, I don't really have any opinion regarding addition of a progressive mode speed override.

No, I still think this should be ini only, because it can mess things up in general, though maybe we can expose at the gui a more sanitize version of it (e.g. allow force 59.94 when it should be 60 or vice verse), but yeah, off topic.

File: pcsx2/Counters.cpp (hunk 1 above)
Comment (RazielZnot) [inline pcsx2/Counters.cpp]: > HDTV 480p -> Vertical frequency of 59.94Hz and a frame frequency of 29.97 fps
> Other progressive modes -> Vertical frequency of 60Hz and a frame frequency of 30 fps

@ssakash you've made mistake, in all **progressive** modes vertical frequency is equal to frame frequency.

Comment (ssakash) [inline pcsx2/Counters.cpp]: Nice catch, I got a bit confused with the logic on counters.

```
u64 Frame = ((u64)PS2CLK * 1000000ULL) / (framesPerSecond * 100).ToIntRounded();
```

**Comment by avih** [comment]
The new commits look good to me. Can I merge them? or are there any other pending updates here?

**Comment by ssakash** [comment]
No more changes left to do, the intended changes have been done.

**Comment by avih** [comment]
Huh.. I must have missed the "merged" note. Is there a reason to keep it open then? If no, then let's close it.

**Comment by ssakash** [comment]
I mean, the changes are done on this PR. They haven't been merged into the master branch yet ;)

**Comment by avih** [comment]
Oh, the merge was for another PR.

So back to square 1

> The new commits look good to me. Can I merge them? or are there any other pending updates here?

Shall I merge?

**Comment by ssakash** [comment]
> Shall I merge?

Yes, feel free to merge it. The PR should also fix the issue mentioned [over here](http://forums.pcsx2.net/Thread-Abnormal-Speed-fps-seen-with-latest-Dev-Builds)

However there's still one more issue,

* If the older save state was saved in PAL video mode then it'll have a vertical frequency limit of ``60Hz`` instead of ``50Hz``. Incrementing the save state version is the only clean way to prevent the issue.

**Comment by avih** [comment]
> Yes, feel free to merge it. The PR should also fix the issue mentioned over here

I _think_ it's the same issue we mentioned earlier here, that loading an older savestate displays incorrect speed, isn't it?



> If the older save state was saved in PAL video mode then it'll have a vertical frequency limit of 60Hz instead of 50Hz. Incrementing the save state version is the only clean way to prevent the issue.

Right, this is another issue then where it actually plays at an incorrect speed, but as far as I understand, it won't happen with new saved states, right?

OK, I'll merge.

As for whether or not to bump the savestate version, I tend not to, because IMO the value of being able to load a saved state is quite high, and even if it displays an incorrect speed or even runs at an incorrect speed, then user can do an in-game save, boot, and load the in-game save, and savestate again - which would fix the issue.

But I don't feel very strongly either way about the version bump.

**Comment by ssakash** [comment]
> I think it's the same issue we mentioned earlier here, that loading an older savestate displays incorrect speed, isn't it?

Yeah

> Right, this is another issue then where it actually plays at an incorrect speed, but as far as I understand, it won't happen with new saved states, right?

Yes, it won't happen on new saved states. (after implementation of ``GS_VideoMode``)

About save state bump - I agree, better to avoid bumping it for trivial issues like these.
